enterprise/backend/src/metabase_enterprise/database_replication/api.clj

```diff
@@ -31,44 +31,55 @@
-                         -1)
-        include-patterns (->> schemas
-                              (filter (comp #{"include"} :type))
-                              (map (comp re-pattern :pattern)))
-        exclude-patterns (->> schemas
-                              (filter (comp #{"exclude"} :type))
-                              (map (comp re-pattern :pattern)))
-        total-row-count (or
-                         (some->>
-                          (t2/hydrate database [:tables :fields])
-                          :tables
-                          (filter (comp (fn [x] (re-matches #"^[A-Za-z0-9_]+$" x)) :name))       ; sanitized name
-                          (filter (fn [t] (some (comp #{:type/PK} :semantic_type) (:fields t)))) ; has-pkey
-                          (filter (fn [{:keys [schema]}]
-                                    (cond
-                                      (empty? schemas)
-                                      true
-                                      (not-empty include-patterns)
-                                      (some #(re-matches % schema) include-patterns)
-                                      (not-empty exclude-patterns)
-                                      (not (some #(re-matches % schema) exclude-patterns)))))
-                          (map :estimated_row_count)
-                          ;; FIXME: `estimated_row_count` might be `nil`, in which case we should tell the user:
-                          ;;  "we don't know yet whether this will work, wait or try your luck"
-                          (reduce +))
-                         0)]
-    (log/infof "Quota left: %s. Estimate db row count: %s" free-quota total-row-count)
-    (< total-row-count free-quota)))
+  (let [free-quota                (or
+                                   (some->> (m/find-first (comp #{"clickhouse-dwh"} :hosting-feature) (quotas))
+                                            ((juxt :soft-limit :usage))
+                                            (apply -))
+                                   -1)
+        include-patterns          (->> schemas
+                                       (filter (comp #{"include"} :type))
+                                       (map (comp re-pattern :pattern)))
+        exclude-patterns          (->> schemas
+                                       (filter (comp #{"exclude"} :type))
+                                       (map (comp re-pattern :pattern)))
+        total-estimated-row-count (or
+                                   (some->>
+                                    (t2/hydrate database [:tables :fields])
+                                    :tables
+                                    (filter (comp (fn [x] (re-matches #"^[A-Za-z0-9_]+$" x)) :name))       ; sanitized name
+                                    (filter (fn [t] (some (comp #{:type/PK} :semantic_type) (:fields t)))) ; has-pkey
+                                    (filter (fn [{:keys [schema]}]
```

Comment:
Should we add the sets of tables that get removed here?  (I'd add three sets / lists and make each complete, i.e. the sets IMO don't have to be disjunct, because that might be easier for the user to understand.)

Reply:
we can. on a need basis. we don't have to think about them before they are required.  But yes, by all means, this map is open for growth